README.md
### project

When submitting jobs we will prepend the jobname with the project name

# NAME

HPC::Runner::Command - Create composable bioinformatics hpc analyses.

# SYNOPSIS

To create a new project

    hpcrunner.pl new MyNewProject

To submit jobs to a cluster

    hpcrunner.pl submit_jobs --infile my_submission.sh

To run jobs on an interactive queue or workstation

    hpcrunner.pl single_node --infile my_submission.sh

# DESCRIPTION

HPC::Runner::Command is a set of libraries for scaffolding data analysis projects,
submitting and executing jobs on an HPC cluster or workstation, and obsessively
logging results.

Get help by heading on over to github and raising an issue. [GitHub ](https://metacpan.org/pod/
https:#github.com-biosails-HPC-Runner-Command-issues).

Please see the complete documentation at [HPC::Runner::Command GitBooks ](https://metacpan.org/pod/
https:#biosails.gitbooks.io-hpc-runner-command-docs-content).

# Quick Start - Create a New Project

You can create a new project, with a sane directory structure by using

        hpcrunner.pl new MyNewProject

# Quick Start - Submit Workflows

## Simple Example

Our simplest example is a single job type with no dependencies - each task is
independent of all other tasks.

### Workflow file

        #preprocess.sh

        echo "preprocess" && sleep 10;
        echo "preprocess" && sleep 10;
        echo "preprocess" && sleep 10;

### Submit to the scheduler

        hpcrunner.pl submit_jobs --infile preprocess.sh

### Look at results!

        tree hpc-runner

### Audit your results

    hpcrunner.pl stats -h
    hpcrunner.pl stats

## Job Type Dependencency Declaration

Most of the time we have jobs that depend upon other jobs.

### Workflow file

        #blastx.sh

        #HPC jobname=unzip
        unzip Sample1.zip
        unzip Sample2.zip
        unzip Sample3.zip

        #HPC jobname=blastx
        #HPC deps=unzip
        blastx --db env_nr --sample Sample1.fasta
        blastx --db env_nr --sample Sample2.fasta
        blastx --db env_nr --sample Sample3.fasta

### Submit to the scheduler

        hpcrunner.pl submit_jobs --infile preprocess.sh

### Look at results!

        tree hpc-runner

## Task Dependencency Declaration

Within a job type we can declare dependencies on particular tasks.

### Workflow file

        #blastx.sh

        #HPC jobname=unzip
        #TASK tags=Sample1
        unzip Sample1.zip
        #TASK tags=Sample2
        unzip Sample2.zip
        #TASK tags=Sample3
        unzip Sample3.zip

        #HPC jobname=blastx
        #HPC deps=unzip
        #TASK tags=Sample1
        blastx --db env_nr --sample Sample1.fasta
        #TASK tags=Sample2
        blastx --db env_nr --sample Sample2.fasta
        #TASK tags=Sample3
        blastx --db env_nr --sample Sample3.fasta

### Submit to the scheduler

        hpcrunner.pl submit_jobs --infile preprocess.sh

### Look at results!

        tree hpc-runner

### Audit your results

    hpcrunner.pl stats -h
    hpcrunner.pl stats

## Declare Scheduler Variables

Each scheduler has its own set of variables. HPC::Runner::Command has a set of
generalized variables for declaring types across templates. For more information
please see [Job Scheduler
Comparison](https://biosails.gitbooks.io/hpc-runner-command-docs/content/job_submission/comparison.html)

Additionally, for workflows with a large number of tasks, please see
[Considerations for Workflows with a Large Number of
Tasks](https://biosails.gitbooks.io/hpc-runner-command-docs/content/design_workflow.html#considerations-for-workflows-with-a-large-number-of-tasks)
for information on how to group tasks together.

### Workflow file

          #blastx.sh

          #HPC jobname=unzip
          #HPC cpus_per_task=1
          #HPC partition=serial
          #HPC commands_per_node=1
    #HPC mem=4GB
          #TASK tags=Sample1
          unzip Sample1.zip
          #TASK tags=Sample2
          unzip Sample2.zip
          #TASK tags=Sample3
          unzip Sample3.zip

          #HPC jobname=blastx
          #HPC cpus_per_task=6
          #HPC deps=unzip
          #TASK tags=Sample1
          blastx --threads 6 --db env_nr --sample Sample1.fasta
          #TASK tags=Sample2
          blastx --threads 6 --db env_nr --sample Sample2.fasta
          #TASK tags=Sample3
          blastx --threads 6 --db env_nr --sample Sample3.fasta

### Submit to the scheduler

        hpcrunner.pl submit_jobs --infile preprocess.sh

### Look at results!

        tree hpc-runner

### Audit your results

    hpcrunner.pl stats -h
    hpcrunner.pl stats

# AUTHOR

Jillian Rowe <jillian.e.rowe@gmail.com>

# Previous Release

This software was previously released under [HPC::Runner](https://metacpan.org/pod/HPC::Runner).
[HPC::Runner::Command](https://metacpan.org/pod/HPC::Runner::Command) is a complete rewrite of the existing library. While it
is meant to have much of the same functionality, it is not backwords compatible.

# Acknowledgements

As of Version 2.41:

This modules continuing development is supported by NYU Abu Dhabi in the Center
for Genomics and Systems Biology. With approval from NYUAD, this information was
generalized and put on github, for which the authors would like to express
their gratitude.

Before Version 2.41

This module was originally developed at and for Weill Cornell Medical College in
Qatar within ITS Advanced Computing Team. With approval from WCMC-Q, this
information was generalized and put on github, for which the authors would like
to express their gratitude.

# COPYRIGHT

Copyright 2016- Jillian Rowe

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Revision history for HPC-Runner-Command

{{$NEXT}}

3.2.14  2018-01-22 10:51:11 GST
	- Pinned List::MoreUtils 0.428
	- Pinned MooseX::App 1.39

3.2.13  2018-01-21 11:52:41 GST
	- Cleaned up dependencies

3.2.12  2018-01-10 14:18:45 GST
	- added back in stats command

3.2.11  2017-09-11 13:50:17 GST
	- Milestones - https://github.com/biosails/HPC-Runner-Command/milestone/3

3.2.10  2017-08-28 09:33:43 GST
	- Rolling back JSON::Logger features for now - problematic on some filesystems
	- Rolling back memory profiling - requires JSON::Logger

3.2.9  2017-08-25 05:16:44 GST
	- Additional fixes for JSON logger
	- Replaced Archive::Tar with Archive::Tar::Wrapper

3.2.8  2017-07-16 17:42:15 EDT
	- Fixed JSON logger - extra check for proc file dissapearing

3.2.7  2017-07-07 11:24:22 GST
	- Bug fix for slurm array deps
	- Added lock file to JSON-Logger. Files in archive are not thread safe.
	- Added experimental tag to JSON logger.

3.2.6  2017-07-05 11:46:45 GST
	- Enhancements to the logging interface
	- New JSON logging interface
	- hpcrunner.pl stats command - works from tar archive and is meant to be a starting point for additional loggers
	- Fixed file paths so tests should run platform independent

3.2.5  2017-06-11 15:53:59 GST
	- Minor cleanup to config files - reload to get any options that may be in plugins

3.2.4  2017-06-07 18:29:58 GST
	- Refactored Scheduler.pm into several more modules
	- Bug fix for use_arrays and commands_per_node
	- Better test coverage for execute_job

3.2.3  2017-05-31 16:45:39 GST
	- When a job fails - only jobs that depend upon that job will not be submitted, but the rest of the workflow is fine

3.2.2  2017-05-29 14:50:37 GST
	- Bug fix for array task dependencies

3.2.1  2017-05-28 15:37:39 GST
	- Bug fix with counting 0th command and single-node mode

3.2.0  2017-05-24 10:26:38 GST
	- Upgrading to minor version

3.1.8  2017-05-24 09:30:51 GST
	- Major API changes for execute_job and execute_array

3.1.7  2017-05-18 10:27:10 GST
	- Hot fix - updated SLURM for new dep array

3.1.6  2017-05-17 15:22:15 GST
	- Made some changes for speed

3.1.5  2017-05-01 09:01:42 GST
	- More logging info

3.1.4  2017-04-30 11:33:17 GST
	- Minor bug fix for plugins

3.1.3  2017-04-30 11:08:50 GST
	- Additional logging info on jobs - hostname, job scheduler id, array id

3.1.2  2017-04-01 11:22:06 GST
	- Made some fixes to environmental variables for slurm
	- Changed duration from %year %month ... to just %day %hours %minutes

3.1.1  2017-03-12 14:44:10 GST
	- Bug fix for ManyConfigs

3.1.0  2017-03-12 11:16:13 GST
	- Official stable release

3.0.1  2017-01-11 10:54:34 GST
	- Bug fixes for cpan tests

3.0.0  2017-01-05 12:26:17 GST
        - This is the beta release of HPC::Runner::Command


lib/HPC/Runner/Command.pm
package HPC::Runner::Command;

use MooseX::App 1.39 qw(Color);

with 'BioSAILs::Utils::Plugin';
with 'BioSAILs::Utils::LoadConfigs';

use MooseX::Types::Path::Tiny qw/Path Paths AbsPath AbsFile/;

app_strict 0;
app_exclude(
    'HPC::Runner::Command::Utils',
    'HPC::Runner::Command::Logger',
    'HPC::Runner::Command::submit_jobs::Utils',
    'HPC::Runner::Command::submit_jobs::Plugin',
    'HPC::Runner::Command::submit_jobs::Logger',
    'HPC::Runner::Command::stats::Logger',
    'HPC::Runner::Command::execute_job::Utils',
    'HPC::Runner::Command::execute_job::Logger',
    'HPC::Runner::Command::execute_job::Base',
);

option '+config_base' => ( default => '.hpcrunner', );

=head3 project

When submitting jobs we will prepend the jobname with the project name

=cut

option 'project' => (
    is            => 'rw',
    isa           => 'Str',
    documentation => 'Give your jobnames an additional project name. '
      . '#HPC jobname=gzip will be submitted as 001_project_gzip',
    required    => 0,
    predicate   => 'has_project',
    cmd_aliases => ['pr'],
);

option 'no_log_json' => (
    traits        => ['Bool'],
    is            => 'rw',
    isa           => 'Bool',
    default       => 0,
    documentation => 'Opt out of writing the tar archive of JSON stats. '
      . 'This may be desirable for especially large workflows.',
);

has 'submission_uuid' => (
    is        => 'rw',
    isa       => 'Str',
    required  => 0,
    predicate => 'has_submissions_uuid',
);

our $VERSION = '3.2.13';

app_strict 0;

sub BUILD { }

=encoding utf-8

=head1 NAME

=begin HTML

<p><img
src="http://github.com/biosails/HPC-Runner-Command/blob/master/_docs/images/rabbit.jpeg"
width="500" height="250" alt="HPC::Runner::Command" /></p>

=end HTML

HPC::Runner::Command - Create composable bioinformatics hpc analyses.

=head1 SYNOPSIS

To create a new project

    hpcrunner.pl new MyNewProject

To submit jobs to a cluster

    hpcrunner.pl submit_jobs --infile my_submission.sh

To run jobs on an interactive queue or workstation

    hpcrunner.pl single_node --infile my_submission.sh

=head1 DESCRIPTION

HPC::Runner::Command is a set of libraries for scaffolding data analysis projects,
submitting and executing jobs on an HPC cluster or workstation, and obsessively
logging results.

Get help by heading on over to github and raising an issue. L<GitHub |
https://github.com/biosails/HPC-Runner-Command/issues>.

Please see the complete documentation at L<HPC::Runner::Command GitBooks |
https://biosails.gitbooks.io/hpc-runner-command-docs/content/>.

=head1 Quick Start - Create a New Project

You can create a new project, with a sane directory structure by using

	hpcrunner.pl new MyNewProject

=head1 Quick Start - Submit Workflows

=head2 Simple Example

Our simplest example is a single job type with no dependencies - each task is
independent of all other tasks.

=head3 Workflow file

	#preprocess.sh

	echo "preprocess" && sleep 10;
	echo "preprocess" && sleep 10;
	echo "preprocess" && sleep 10;

=head3 Submit to the scheduler

	hpcrunner.pl submit_jobs --infile preprocess.sh

=head3 Look at results!

	tree hpc-runner

=head3 Audit your results

  hpcrunner.pl stats -h
  hpcrunner.pl stats

=head2 Job Type Dependencency Declaration

Most of the time we have jobs that depend upon other jobs.

=head3 Workflow file

	#blastx.sh

	#HPC jobname=unzip
	unzip Sample1.zip
	unzip Sample2.zip
	unzip Sample3.zip

	#HPC jobname=blastx
	#HPC deps=unzip
	blastx --db env_nr --sample Sample1.fasta
	blastx --db env_nr --sample Sample2.fasta
	blastx --db env_nr --sample Sample3.fasta

=head3 Submit to the scheduler

	hpcrunner.pl submit_jobs --infile preprocess.sh

=head3 Look at results!

	tree hpc-runner

=head2 Task Dependencency Declaration

Within a job type we can declare dependencies on particular tasks.

=head3 Workflow file

	#blastx.sh

	#HPC jobname=unzip
	#TASK tags=Sample1
	unzip Sample1.zip
	#TASK tags=Sample2
	unzip Sample2.zip
	#TASK tags=Sample3
	unzip Sample3.zip

	#HPC jobname=blastx
	#HPC deps=unzip
	#TASK tags=Sample1
	blastx --db env_nr --sample Sample1.fasta
	#TASK tags=Sample2
	blastx --db env_nr --sample Sample2.fasta
	#TASK tags=Sample3
	blastx --db env_nr --sample Sample3.fasta

=head3 Submit to the scheduler

	hpcrunner.pl submit_jobs --infile preprocess.sh

=head3 Look at results!

	tree hpc-runner

=head3 Audit your results

  hpcrunner.pl stats -h
  hpcrunner.pl stats

=cut

=head2 Declare Scheduler Variables

Each scheduler has its own set of variables. HPC::Runner::Command has a set of
generalized variables for declaring types across templates. For more information
please see L<Job Scheduler
Comparison|https://biosails.gitbooks.io/hpc-runner-command-docs/content/job_submission/comparison.html>

Additionally, for workflows with a large number of tasks, please see
L<Considerations for Workflows with a Large Number of
Tasks|https://biosails.gitbooks.io/hpc-runner-command-docs/content/design_workflow.html#considerations-for-workflows-with-a-large-number-of-tasks>
for information on how to group tasks together.

=head3 Workflow file

	#blastx.sh

	#HPC jobname=unzip
	#HPC cpus_per_task=1
	#HPC partition=serial
	#HPC commands_per_node=1
  #HPC mem=4GB
	#TASK tags=Sample1
	unzip Sample1.zip
	#TASK tags=Sample2
	unzip Sample2.zip
	#TASK tags=Sample3
	unzip Sample3.zip

	#HPC jobname=blastx
	#HPC cpus_per_task=6
	#HPC deps=unzip
	#TASK tags=Sample1
	blastx --threads 6 --db env_nr --sample Sample1.fasta
	#TASK tags=Sample2
	blastx --threads 6 --db env_nr --sample Sample2.fasta
	#TASK tags=Sample3
	blastx --threads 6 --db env_nr --sample Sample3.fasta

=head3 Submit to the scheduler

	hpcrunner.pl submit_jobs --infile preprocess.sh

=head3 Look at results!

	tree hpc-runner

=head3 Audit your results

  hpcrunner.pl stats -h
  hpcrunner.pl stats

=cut

__PACKAGE__->meta()->make_immutable();

1;

__END__

=head1 AUTHOR

Jillian Rowe E<lt>jillian.e.rowe@gmail.comE<gt>

=head1 Previous Release

This software was previously released under L<HPC::Runner>.
L<HPC::Runner::Command> is a complete rewrite of the existing library. While it
is meant to have much of the same functionality, it is not backwords compatible.

=head1 Acknowledgements

As of Version 2.41:

This modules continuing development is supported by NYU Abu Dhabi in the Center
for Genomics and Systems Biology. With approval from NYUAD, this information was
generalized and put on github, for which the authors would like to express
their gratitude.

Before Version 2.41

This module was originally developed at and for Weill Cornell Medical College in
Qatar within ITS Advanced Computing Team. With approval from WCMC-Q, this
information was generalized and put on github, for which the authors would like
to express their gratitude.

=head1 COPYRIGHT

Copyright 2016- Jillian Rowe

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
